### PR TITLE
Grain Velocity Computation

### LINKED ISSUE #14881: Grain Velocity #14880
### **Reason**

Visualization of velocity inside grain structures will be useful to see the movement of grains inside of the simulation. 

### **Design**

Implementation in Auxkernel seems to be the best option for usability. Using this equation for calculation: 
1/abs(gradη)×(δ η)/(δ t)
and reducing the numerical instability by using eta around the edges seems to be a good solution. 
Also testing using findValueOnLine is useful for determining if the correct values are returned. 

### **Impact**

Increased visualization of grain boundary properties.

### PR DESCRIPTION
**Pull request by killerpenguin12**

### CODE REVIEW

**Comment by dschwen** [comment]
"Add files via upload" is not an acceptable commit message.

File: modules/phase_field/include/auxkernels/calcVelocityAux.h [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+#include "AuxKernel.h"
+//#include "DerivativeMaterialInterface.h"
+
+class calcVelocityAux;
+class GrainTracker;
+
+template <>
+InputParameters validParams<calcVelocityAux>();
+
+class calcVelocityAux : public AuxKernel
```
Comment (dschwen) [inline modules/phase_field/include/auxkernels/calcVelocityAux.h]: We use CamelCase for class names. I suggest renaming to `GrainBoundaryVelocity`

File: modules/phase_field/src/auxkernels/calcVelocityAux.C [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+
+template <>
+InputParameters
+validParams<calcVelocityAux>()
+{
+  InputParameters params = validParams<AuxKernel>();
+  params.addClassDescription(
+      "calculate the velocity of grains while avoiding numerical instability.");
+  params.addParam<Real>("eta_bottom", 0.20, "start point on area of eta we are interested in.");
+  params.addParam<Real>("eta_top", 0.80, "end point on area of eta we are interested in.");
```
Comment (dschwen) [inline modules/phase_field/src/auxkernels/calcVelocityAux.C]: this is not an area but a "range"

File: modules/phase_field/src/auxkernels/calcVelocityAux.C [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+Real
+calcVelocityAux::computeValue()
+{
+  _value = 0;
+  new_val = 0;
+  for (unsigned int i = 0; i < _op_num; ++i)
+  {
+    // To make sure we are getting data from the area we want to reduce numerical instability.
+    if ((*_eta[i])[_qp] > _eta_bottom && (*_eta[i])[_qp] < _eta_top)
+      new_val = (1 / ((*_grad_eta[i])[_qp].norm())) * ((*deta_dt[i])[_qp]);
```
Comment (dschwen) [inline modules/phase_field/src/auxkernels/calcVelocityAux.C]: there are way too many parenthesis in this line

File: modules/phase_field/src/auxkernels/calcVelocityAux.C [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+  return params;
+}
+
+calcVelocityAux::calcVelocityAux(const InputParameters & parameters)
+  : AuxKernel(parameters),
+    _eta_bottom(getParam<Real>("eta_bottom")),
+    _eta_top(getParam<Real>("eta_top")),
+    _op_num(coupledComponents("v")),
+    _eta(_op_num),
+    deta_dt(_op_num),
```
Comment (dschwen) [inline modules/phase_field/src/auxkernels/calcVelocityAux.C]: missing leading underscore

File: modules/phase_field/src/auxkernels/calcVelocityAux.C [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "calcVelocityAux.h"
+//#include "FEProblem.h"
+
+registerMooseObject("PhaseFieldApp", calcVelocityAux);
+
+template <>
+InputParameters
+validParams<calcVelocityAux>()
```
Comment (dschwen) [inline modules/phase_field/src/auxkernels/calcVelocityAux.C]: please use the new static member scheme for validParams

File: modules/phase_field/src/auxkernels/calcVelocityAux.C [hunk 6]
```diff
@@ -0,0 +1,64 @@
...
+//* https://www.mooseframework.org
+//*
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#include "calcVelocityAux.h"
+//#include "FEProblem.h"
```
Comment (dschwen) [inline modules/phase_field/src/auxkernels/calcVelocityAux.C]: remove

File: modules/phase_field/include/auxkernels/calcVelocityAux.h [hunk 7]
```diff
@@ -0,0 +1,39 @@
...
+//* All rights reserved, see COPYRIGHT for full restrictions
+//* https://github.com/idaholab/moose/blob/master/COPYRIGHT
+//*
+//* Licensed under LGPL 2.1, please see LICENSE for details
+//* https://www.gnu.org/licenses/lgpl-2.1.html
+
+#pragma once
+
+#include "AuxKernel.h"
+//#include "DerivativeMaterialInterface.h"
```
Comment (dschwen) [inline modules/phase_field/include/auxkernels/calcVelocityAux.h]: remove

File: modules/phase_field/src/auxkernels/calcVelocityAux.C [hunk 8]
```diff
@@ -0,0 +1,64 @@
...
+
+registerMooseObject("PhaseFieldApp", calcVelocityAux);
+
+template <>
+InputParameters
+validParams<calcVelocityAux>()
+{
+  InputParameters params = validParams<AuxKernel>();
+  params.addClassDescription(
+      "calculate the velocity of grains while avoiding numerical instability.");
```
Comment (dschwen) [inline modules/phase_field/src/auxkernels/calcVelocityAux.C]: The "while avoiding numerical instability" is not a particularly useful bit of information and this is rather computing the velocity of *grain boundaries*.

**Review by dschwen — CHANGES_REQUESTED** [review]

File: modules/phase_field/.atom/.apmrc [hunk 9]
```diff
@@ -0,0 +1 @@
+strict-ssl=false
```
Comment (dschwen) [inline modules/phase_field/.atom/.apmrc]: remove this file

File: modules/phase_field/doc/content/source/auxkernels/calcVelocityAux.md [hunk 10]
```diff
@@ -0,0 +1 @@
+!template load file=stubs/moose_object.md.template name=calcVelocity syntax=/AuxKernels/calcVelocityAux
```
Comment (dschwen) [inline modules/phase_field/doc/content/source/auxkernels/calcVelocityAux.md]: remove this file

File: modules/phase_field/doc/content/source/auxkernels/GrainBoundaryVelocity.md [hunk 11]
```diff
@@ -0,0 +1 @@
+!template load file=stubs/moose_object.md.template name=GrainBoundaryVelocity syntax=/AuxKernels/GrainBoundaryVelocity
```
Comment (dschwen) [inline modules/phase_field/doc/content/source/auxkernels/GrainBoundaryVelocity.md]: Add actual documentation here!

File: modules/phase_field/src/auxkernels/GrainBoundaryVelocity.C [hunk 12]
```diff
@@ -0,0 +1,61 @@
...
+
+GrainBoundaryVelocity::GrainBoundaryVelocity(const InputParameters & parameters)
+  : AuxKernel(parameters),
+    _eta_bottom(getParam<Real>("eta_bottom")),
+    _eta_top(getParam<Real>("eta_top")),
+    _op_num(coupledComponents("v")),
+    _eta(_op_num),
+    _deta_dt(_op_num),
+    _grad_eta(_op_num)
+
```
Comment (dschwen) [inline modules/phase_field/src/auxkernels/GrainBoundaryVelocity.C]: remove blank line

File: modules/phase_field/src/auxkernels/GrainBoundaryVelocity.C [hunk 13]
```diff
@@ -0,0 +1,61 @@
...
+Real
+GrainBoundaryVelocity::computeValue()
+{
+  _value = 0;
+  _new_val = 0;
+  for (unsigned int i = 0; i < _op_num; ++i)
+  {
+    // To make sure we are getting data from the range that we want
+    if ((*_eta[i])[_qp] > _eta_bottom && (*_eta[i])[_qp] < _eta_top)
+      _new_val = 1 / (*_grad_eta[i])[_qp].norm() * (*_deta_dt[i])[_qp];
```
Comment (dschwen) [inline modules/phase_field/src/auxkernels/GrainBoundaryVelocity.C]: Please distinguish whole floating point numbers from integers by adding `.0` (this is important in a small number of cases, but it's best to do it consistently)

File: modules/phase_field/test/tests/Grain_Velocity_Computation/GrainBoundaryVelocityTest.i [hunk 14]
```diff
@@ -0,0 +1,137 @@
...
+  dim = 2
+  nx = 10
+  ny = 10
+  nz = 0
+  xmin = 0
+  xmax = 1000
+  ymin = 0
+  ymax = 1000
+  zmin = 0
+  zmax = 0
```
Comment (dschwen) [inline modules/phase_field/test/tests/Grain_Velocity_Computation/GrainBoundaryVelocityTest.i]: remove all the z stuff, it is unnecessary

File: modules/phase_field/test/tests/Grain_Velocity_Computation/GrainBoundaryVelocityTest.i [hunk 15]
```diff
@@ -0,0 +1,137 @@
...
+      invalue = 1.0
+      outvalue = 0.0
+    [../]
+  [../]
+[]
+
+[AuxVariables]
+  [./velocity]
+  order = CONSTANT
+  family = MONOMIAL
```
Comment (dschwen) [inline modules/phase_field/test/tests/Grain_Velocity_Computation/GrainBoundaryVelocityTest.i]: indent

File: modules/phase_field/test/tests/Grain_Velocity_Computation/GrainBoundaryVelocityTest.i [hunk 16]
```diff
@@ -0,0 +1,137 @@
...
+[GlobalParams]
+  op_num = 2
+  var_name_base = gr
+  v = 'gr0 gr1'
+[]
+
+[Variables]
+  [./gr0]
+    order = FIRST
+    family = LAGRANGE
```
Comment (dschwen) [inline modules/phase_field/test/tests/Grain_Velocity_Computation/GrainBoundaryVelocityTest.i]: you can leave out family and order, they default to the values you're setting. Shorter files are more readable.

File: modules/phase_field/test/tests/Grain_Velocity_Computation/GrainBoundaryVelocityTest.i [hunk 17]
```diff
@@ -0,0 +1,137 @@
...
+
+[AuxKernels]
+  [./velocity]
+    type = GrainBoundaryVelocity
+    variable = velocity
+  [../]
+[]
+
+[BCs]
+  active = ' '
```
Comment (dschwen) [inline modules/phase_field/test/tests/Grain_Velocity_Computation/GrainBoundaryVelocityTest.i]: if none of those are active I'd prefer to just remove the entire block

File: modules/phase_field/test/tests/Grain_Velocity_Computation/GrainBoundaryVelocityTest.i [hunk 18]
```diff
@@ -0,0 +1,137 @@
...
+
+  [./Periodic]
+    [./left_right]
+      primary = 0
+      secondary = 2
+      translation = '0 1000 0'
```
Comment (dschwen) [inline modules/phase_field/test/tests/Grain_Velocity_Computation/GrainBoundaryVelocityTest.i]: use `auto_direction` for periodicity with generated meshes!

File: modules/phase_field/test/tests/Grain_Velocity_Computation/GrainBoundaryVelocityTest.i [hunk 19]
```diff
@@ -0,0 +1,137 @@
...
+    GBmob0 = 2.5e-6 #m^4/(Js) from Schoenfelder 1997
+    Q = 0.23 #Migration energy in eV
+    GBenergy = 0.708 #GB energy in J/m^2
+  [../]
+[]
+
+[Preconditioning]
+  [./SMP]
+   type = SMP
+   full = true
```
Comment (dschwen) [inline modules/phase_field/test/tests/Grain_Velocity_Computation/GrainBoundaryVelocityTest.i]: fix indentaton

File: modules/phase_field/test/tests/Grain_Velocity_Computation/GrainBoundaryVelocityTest.i [hunk 20]
```diff
@@ -0,0 +1,137 @@
...
+  petsc_options_value = 'hypre boomeramg 31'
+
+  l_tol = 1.0e-4
+  l_max_its = 30
+
+  nl_max_its = 20
+  nl_rel_tol = 1.0e-9
+
+  start_time = 0.0
+  num_steps = 5
```
Comment (dschwen) [inline modules/phase_field/test/tests/Grain_Velocity_Computation/GrainBoundaryVelocityTest.i]: This test is very likely to be too heavy. If it takes more than **two seconds** in opt it needs to get simplified.

**Review by dschwen — CHANGES_REQUESTED** [review]
See comments. You're also missing a `tests` file, so your test won't actually get run. Also, you added 3 gold files, which are two too many. My suggestion would be to drop adaptivity and uniform_refine. Just go with a simple 20x20 mesh.

File: modules/phase_field/doc/content/source/auxkernels/GrainBoundaryVelocity.md [hunk 21]
```diff
@@ -0,0 +1,28 @@
...
+we are using the equation:
+\begin{equation}
+\frac{1}{\left | \triangledown \eta \right |} \frac{d\eta}{dt} = v
+\end{equation}
+which computes the perpendicular velocity at each quadrature point in the range
+specified.
+
+## Example Input File Syntax
+
+This AuxKernel only requires an AuxVariable to compute on.
```
Comment (dschwen) [inline modules/phase_field/doc/content/source/auxkernels/GrainBoundaryVelocity.md:18]: Not true, it also required order parameters... I'd just remove that line

File: modules/phase_field/test/tests/Grain_Velocity_Computation/tests [hunk 22]
```diff
@@ -0,0 +1,7 @@
+[Tests]
+  [./GrainBoundaryVelocityTest]
+    type = 'Exodiff'
+    input = 'GrainBoundaryVelocityTest.i'
+    exodiff = 'GrainBoundaryVelocityTest_out.e'
+  [../]
```
Comment (dschwen) [inline modules/phase_field/test/tests/Grain_Velocity_Computation/tests:6]: This needs `design` `requirements` `issues`

**Review by dschwen — COMMENTED** [review]
Almost ready. Please squash your commits.

**Comment by dschwen** [comment]
You have a merge commit in your branch now :-/

**Comment by dschwen** [comment]
Do

```
git reset --hard  66cbb62
git pull --rebase upstream devel
```

To remove it and properly update your branch.
Then use

 ```
git rebase -i HEAD~19
```

to squash your commits (you may need to Google some instructions if you've never done this).
Then push your changes with 

```
git push -f
```

File: modules/phase_field/test/tests/Grain_Velocity_Computation/tests [hunk 23]
```diff
@@ -0,0 +1,11 @@
+[Tests]
+  [./GrainBoundaryVelocityTest]
+    type = 'Exodiff'
+    input = 'GrainBoundaryVelocityTest.i'
+    exodiff = 'GrainBoundaryVelocityTest_out.e'
+
+    issues = '#14887'
+    design = 'GrainBoundaryVelocity.md'
+    requirement = 'We will be able to visualize the velocity in the simulation.'
```
Comment (permcody) [inline modules/phase_field/test/tests/Grain_Velocity_Computation/tests:9]:
```suggestion
    requirement = 'The system shall be able to provide a velocity vector field indicating grain boundary movement for visualization purposes.'
```

**Review by permcody — DISMISSED** [review]

File: modules/phase_field/src/auxkernels/GrainBoundaryVelocity.C [hunk 24]
```diff
@@ -0,0 +1,60 @@
...
+    _deta_dt[i] = &coupledDot("v", i);
+    _eta[i] = &coupledValue("v", i);
+  }
+}
+
+Real
+GrainBoundaryVelocity::computeValue()
+{
+  _value = 0.0;
+  _new_val = 0.0;
```
Comment (dschwen) [inline modules/phase_field/src/auxkernels/GrainBoundaryVelocity.C:48]: These two variables should not be member variables. But we can fix that later.

**Review by dschwen — APPROVED** [review]

**Pull request merged**